bridge-ffi/src/lib.rs

```diff
@@ -0,0 +1,650 @@
+        // Mimics Vec::into_raw_parts that's only available in nightly
+        let mut vec = std::mem::ManuallyDrop::new(vec);
+        tmprl_bytes_t {
+            bytes: vec.as_mut_ptr(),
+            len: vec.len(),
+            cap: vec.capacity(),
+            disable_free: false,
+        }
+    }
+
+    fn from_vec_disable_free(vec: Vec<u8>) -> tmprl_bytes_t {
+        let mut b = tmprl_bytes_t::from_vec(vec);
+        b.disable_free = true;
+        b
+    }
+
+    fn into_raw(self) -> *mut tmprl_bytes_t {
+        Box::into_raw(Box::new(self))
+    }
+}
+
+// We only impl this because they are required by lazy_static and raw pointers
+// are not usually safe for send
+unsafe impl Send for tmprl_bytes_t {}
+unsafe impl Sync for tmprl_bytes_t {}
+
+impl Drop for tmprl_bytes_t {
+    fn drop(&mut self) {
+        // In cases where freeing is disabled (or technically some other
+        // drop-but-not-freed situation though we don't expect any), the bytes
+        // remain non-null so we re-own them here
+        if !self.bytes.is_null() {
+            unsafe { Vec::from_raw_parts(self.bytes as *mut u8, self.len, self.cap) };
+        }
+    }
+}
+
+/// Free a set of bytes. The first parameter can be null in cases where a
+/// tmprl_core_t instance isn't available. If the second parameter is null, this
+/// is a no-op.
+#[no_mangle]
+pub extern "C" fn tmprl_bytes_free(core: *mut tmprl_core_t, bytes: *const tmprl_bytes_t) {
+    // Bail if freeing is disabled
+    unsafe {
+        if bytes.is_null() || (*bytes).disable_free {
+            return;
+        }
+    }
+    let bytes = bytes as *mut tmprl_bytes_t;
+    // Return vec back to core before dropping bytes
+    let vec = unsafe { Vec::from_raw_parts((*bytes).bytes as *mut u8, (*bytes).len, (*bytes).cap) };
+    // Set to null so the byte dropper doesn't try to free it
+    unsafe { (*bytes).bytes = std::ptr::null_mut() };
+    // Return only if core is non-null
+    if !core.is_null() {
+        let core = unsafe { &mut *core };
+        core.return_buf(vec);
+    }
+    unsafe {
+        Box::from_raw(bytes);
+    }
+}
+
+// Used for maintaining pointer to user data across threads
+struct UserDataHandle(*mut libc::c_void);
+unsafe impl Send for UserDataHandle {}
+unsafe impl Sync for UserDataHandle {}
+
+lazy_static::lazy_static! {
+    static ref DEFAULT_INIT_RESPONSE_BYTES: tmprl_bytes_t = {
+        tmprl_bytes_t::from_vec_disable_free(bridge::InitResponse::default().encode_to_vec())
+    };
+
+    static ref DEFAULT_SHUTDOWN_RESPONSE_BYTES: tmprl_bytes_t = {
+        tmprl_bytes_t::from_vec_disable_free(bridge::ShutdownResponse::default().encode_to_vec())
+    };
+
+    static ref DEFAULT_REGISTER_WORKER_RESPONSE_BYTES: tmprl_bytes_t = {
+        tmprl_bytes_t::from_vec_disable_free(bridge::RegisterWorkerResponse::default().encode_to_vec())
+    };
+}
+
+/// A runtime owned by Core. This must be passed to tmprl_runtime_free when no
```

Comment:
Technically not owned by `Core`, `Core` is spawned in a runtime. We could be more accurate in the terminology here.

Reply:
To clarify, I use the capitalized `Core` to refer to this whole static library (so it's owned by the library), and I use the lowercased `core` to refer to an instance. This statement is saying the caller does not own the runtime, this library does.